catalog/transformer/catalog-transformer-csv-common/src/main/java/ddf/catalog/transformer/csv/common/CsvTransformer.java

```diff
@@ -177,4 +182,50 @@ public static Set<AttributeDescriptor> getOnlyRequestedAttributes(
 
     return attributes;
   }
+
+  /**
+   * Given a list of {@link Metacard}s, return only the attribute descriptors that have a non-empty
+   * value. Returns a set of {@link AttributeDescriptor}s containing the attributes with non-empty
+   * values.
+   */
+  public static Set<AttributeDescriptor> getNonEmptyValueAttributes(
+      final List<Metacard> metacards) {
+    final Set<AttributeDescriptor> attributes = getAllCsvAttributeDescriptors(metacards);
+
+    Set<AttributeDescriptor> nonEmptyAttributes = new HashSet<>();
+    loop:
+    for (AttributeDescriptor attribute : attributes) {
+      for (Metacard m : metacards) {
+        switch (attribute.getType().getAttributeFormat()) {
+          case STRING:
+          case XML:
+          case GEOMETRY:
+            if (StringUtils.isNotEmpty(m.getAttribute(attribute.getName()).getValue().toString())) {
+              nonEmptyAttributes.add(attribute);
+              continue loop;
+            }
+            break;
+          case INTEGER:
+          case LONG:
+          case DOUBLE:
+          case FLOAT:
+          case SHORT:
+          case DATE:
+          case BOOLEAN:
+            if (Objects.nonNull(m.getAttribute(attribute.getName()).getValue())) {
+              nonEmptyAttributes.add(attribute);
+              continue loop;
+            }
+            break;
+          case BINARY:
+          case OBJECT:
+          default:
+            // do nothing
+            break;
+        }
+      }
+    }
```

Comment:
Check if this is equivalent to what you have:
```java
public static Set<AttributeDescriptor> getNonEmptyValueAttributes(final List<Metacard> metacards) {
    final Set<AttributeDescriptor> attributes = getAllCsvAttributeDescriptors(metacards);
    final Set<AttributeDescriptor> nonEmptyAttributes = new HashSet<>();

    for (final AttributeDescriptor attribute : attributes) {
        final boolean hasNonEmptyValue = metacards.stream().anyMatch(metacard -> isNonEmptyValue(metacard, attribute));
        if (hasNonEmptyValue) {
          nonEmptyAttributes.add(attribute);
        }
    }

    return nonEmptyAttributes;
}

private static boolean isNonEmptyValue(Metacard metacard, AttributeDescriptor descriptor) {
    final Attribute attribute = metacard.getAttribute(descriptor.getName());
    switch (descriptor.getType().getAttributeFormat()) {
        case STRING:
        case XML:
        case GEOMETRY:
          return attribute != null && StringUtils.isNotEmpty((String) attribute.getValue());
        case INTEGER:
        case LONG:
        case DOUBLE:
        case FLOAT:
        case SHORT:
        case DATE:
        case BOOLEAN:
          return attribute != null && attribute.getValue() != null;
        default:
            return false;
    }
}
```

This will reduce nesting and allow us to avoid using a continue with a label. The helper method is basically the same as your existing `switch` statement, but with null checks on the attribute.

Reply:
Thank you - much cleaner than my loop and labels.